File: src/core/safety/permissionService.ts
/**
 * Permission decision engine per spec §4.6.
 *
 * Decision order (deny-first wins):
 *   1. settings.deny    — explicit blocks ALWAYS win, even over a cached allow.
 *   2. auto-approval    — heuristic safe-list (G5.G.6 / G7.5.A.9): read-only
 *                         tools and pure-read Bash commands short-circuit to
 *                         `allow` WITHOUT prompting the user. Deny remains
 *                         above this so the user's explicit veto wins.
 *   3. cache            — session-scoped user approvals (scope=session).
 *   4. settings.allow   — user_global / project_shared / project_local merged tier.
 *   5. otherwise: ask   — emit suggested pattern + available scopes.
 *
 * Compound bash special case: if the command contains `&&`, `||`, or `;`,
 * we force scope=["once"] only. Approving `git status && rm -rf /` for the
 * whole session would let later `git status && <anything>` slip through.
 *
 * G7.5.A.9: auto-approval consultation. The classifier in
 * `auto-approval.ts` already enforces the composition rule (a compound's
 * unsafe leg poisons the whole), so we can safely consult it even for
 * Bash invocations — but we still keep the INV-15 scope clamp below the
 * auto-approval check so an "allow" decision NEVER touches scopes.
 */
import { matchPattern, parsePatternString, type ToolInvocation } from "./patterns";
import { splitCompound } from "./bashCompoundAnalyzer";
import type { MergedSettings } from "./settingsHierarchy";
import { PermissionCache } from "./permissionCache";
import { classifyToolForAutoApproval, type ToolCall } from "./auto-approval";

export type PermissionScope = "once" | "session" | "always";

export type PermissionDecision =
  | { kind: "allow"; reason: "cached" | "settings_allow" | "auto_approved" }
  | { kind: "deny"; reason: "settings_deny" | "compound_unsafe" }
  | { kind: "ask"; suggested_pattern: string; available_scopes: PermissionScope[] };

export interface PermissionServiceOpts {
  cache: PermissionCache;
  settings: MergedSettings;
}

export function check(
  sessionId: string,
  inv: ToolInvocation,
  opts: PermissionServiceOpts,
): PermissionDecision {
  // 0. Compound bash guard (INV-15): always redirect to ask with ["once"].
  //    Must be checked BEFORE allow-list to prevent "always" scope leak.
  let scopes: PermissionScope[] = ["once", "session", "always"];
  let isCompound = false;
  if (inv.tool === "Bash" && typeof inv.input.command === "string") {
    if (splitCompound(inv.input.command).length > 1) {
      scopes = ["once"];
      isCompound = true;
    }
  }

  // 1. Deny first — absolute veto.
  for (const denyStr of opts.settings.deny) {
    const p = parsePatternString(denyStr);
    if (p && matchPattern(p, inv)) {
      return { kind: "deny", reason: "settings_deny" };
    }
  }

  // 2. Auto-approval heuristic (G7.5.A.9). Runs AFTER deny so any explicit
  //    veto from the user still wins. The classifier is default-deny; only
  //    a positive "allow" from it short-circuits, otherwise we fall through
  //    to the regular allow/cache/ask path.
  const autoCall: ToolCall = { tool: inv.tool, input: inv.input };
  const auto = classifyToolForAutoApproval(autoCall);
  if (auto.decision === "allow") {
    return { kind: "allow", reason: "auto_approved" };
  }

  // 3. Session cache (scope=session approvals).
  for (const cached of opts.cache.list(sessionId)) {
    const p = parsePatternString(cached);
    if (p && matchPattern(p, inv)) {
      return { kind: "allow", reason: "cached" };
    }
  }

  // 4. Settings allow (merged tier). Skip for compound bash (INV-15).
  if (!isCompound) {
    for (const allowStr of opts.settings.allow) {
      const p = parsePatternString(allowStr);
      if (p && matchPattern(p, inv)) {
        return { kind: "allow", reason: "settings_allow" };
      }
    }
  }

  // 5. Need to ask the user. Scopes already restricted for compound.
  return {
    kind: "ask",
    suggested_pattern: suggestPattern(inv),
    available_scopes: scopes,
  };
}

function suggestPattern(inv: ToolInvocation): string {
  if (inv.tool === "Bash" && typeof inv.input.command === "string") {
    const first = inv.input.command.trim().split(/\s+/)[0] ?? "";
    return `Bash(${first}:*)`;
  }
  if (inv.tool === "Edit" && typeof inv.input.file_path === "string") {
    return `Edit(${inv.input.file_path})`;
  }
  if (inv.tool === "WebFetch" && typeof inv.input.url === "string") {
    try {
      const u = new URL(inv.input.url);
      return `WebFetch(domain:${u.hostname})`;
    } catch {
      return `WebFetch(*)`;
    }
  }
  return inv.tool;
}

File: crates/apohara-safety/src/permission_service.rs
//! Permission decision engine — ports `src/core/safety/permissionService.ts`.
//!
//! Decision order (deny-first wins):
//!   1. settings.deny    — explicit blocks ALWAYS win, even over a cached allow.
//!   2. auto-approval    — heuristic safe-list short-circuits without prompt.
//!   3. cache            — session-scoped user approvals (scope=session).
//!   4. settings.allow   — merged tier; skipped for compound bash (INV-bash-scope).
//!   5. otherwise: ask   — suggested pattern + available scopes.
//!
//! Compound bash special case: when the command is detected as compound,
//! `available_scopes` is clamped to `[Once]` only and `settings.allow` is
//! skipped, so `git status && rm -rf /` cannot inherit a `Bash(git:*)`
//! always-scope approval.

use crate::auto_approval::{classify_tool_for_auto_approval, AutoApprovalDecision};
use crate::bash_compound::split_compound;
use crate::patterns::{match_pattern, parse_pattern_string, ToolInvocation};
use crate::permission_cache::PermissionCache;
use crate::permission_grid::PermissionScope;
use crate::settings_hierarchy::MergedSettings;
use serde::{Deserialize, Serialize};
use url::Url;

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(tag = "kind", rename_all = "snake_case")]
pub enum PermissionDecision {
    Allow {
        reason: AllowReason,
    },
    Deny {
        reason: DenyReason,
    },
    Ask {
        suggested_pattern: String,
        available_scopes: Vec<PermissionScope>,
    },
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum AllowReason {
    Cached,
    SettingsAllow,
    AutoApproved,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum DenyReason {
    SettingsDeny,
    CompoundUnsafe,
}

pub struct PermissionServiceOpts<'a> {
    pub cache: &'a PermissionCache,
    pub settings: &'a MergedSettings,
}

pub fn check(
    session_id: &str,
    inv: &ToolInvocation,
    opts: PermissionServiceOpts<'_>,
) -> PermissionDecision {
    // 0. Compound bash guard (INV-bash-scope): always redirect to ask with [Once].
    //    Must be checked BEFORE allow-list to prevent "always" scope leak.
    let mut scopes = vec![
        PermissionScope::Once,
        PermissionScope::Session,
        PermissionScope::Always,
    ];
    let mut is_compound = false;
    if inv.tool == "Bash" {
        if let Some(cmd) = inv.input.get("command").and_then(|v| v.as_str()) {
            if split_compound(cmd).len() > 1 {
                scopes = vec![PermissionScope::Once];
                is_compound = true;
            }
        }
    }

    // 1. Deny first — absolute veto.
    for deny_str in &opts.settings.deny {
        if let Some(p) = parse_pattern_string(deny_str) {
            if match_pattern(&p, inv) {
                return PermissionDecision::Deny {
                    reason: DenyReason::SettingsDeny,
                };
            }
        }
    }

    // 2. Auto-approval (runs AFTER deny so explicit user veto still wins).
    if let AutoApprovalDecision::Allow { .. } = classify_tool_for_auto_approval(inv) {
        return PermissionDecision::Allow {
            reason: AllowReason::AutoApproved,
        };
    }

    // 3. Session cache (scope=session approvals).
    for cached in opts.cache.list(session_id) {
        if let Some(p) = parse_pattern_string(&cached) {
            if match_pattern(&p, inv) {
                return PermissionDecision::Allow {
                    reason: AllowReason::Cached,
                };
            }
        }
    }

    // 4. Settings allow (skipped for compound — INV-bash-scope).
    if !is_compound {
        for allow_str in &opts.settings.allow {
            if let Some(p) = parse_pattern_string(allow_str) {
                if match_pattern(&p, inv) {
                    return PermissionDecision::Allow {
                        reason: AllowReason::SettingsAllow,
                    };
                }
            }
        }
    }

    PermissionDecision::Ask {
        suggested_pattern: suggest_pattern(inv),
        available_scopes: scopes,
    }
}

fn suggest_pattern(inv: &ToolInvocation) -> String {
    if inv.tool == "Bash" {
        if let Some(cmd) = inv.input.get("command").and_then(|v| v.as_str()) {
            let first = cmd.split_whitespace().next().unwrap_or("");
            return format!("Bash({first}:*)");
        }
    }
    if inv.tool == "Edit" {
        if let Some(file) = inv.input.get("file_path").and_then(|v| v.as_str()) {
            return format!("Edit({file})");
        }
    }
    if inv.tool == "WebFetch" {
        if let Some(u) = inv.input.get("url").and_then(|v| v.as_str()) {
            return match Url::parse(u).ok().and_then(|p| p.host_str().map(|s| s.to_string())) {
                Some(host) => format!("WebFetch(domain:{host})"),
                None => "WebFetch(*)".to_string(),
            };
        }
    }
    inv.tool.clone()
}
